Public API for defining command line options for tasks

wilkinsona opened the pull request:
### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

In its current form, this PR is a step towards implementing #2402. I don't believe it's complete (for example no documentation updates have been made), but I'm submitting it now for review and, hopefully, confirmation that I'm heading in the right direction.

The title of #2402 suggests making the parsing API public. I've left `OptionReader` as part of the internal API for now as it's only used by other internals. Doing so avoids a cyclic dependency between the `org.gradle.api.internal.tasks.options` and `org.gradle.api.tasks.options` packages that would occur with the current structure if `OptionReader` and `OptionDescriptor` were moved to the latter.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] [Link to Design Spec](https://github.com/gradle/gradle/tree/master/design-docs) for changes that affect more than 1 public API (that is, not in an `internal` package) or updates to > 20 files
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew quickCheck <impacted-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

bmuschko commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/AbstractOptionElement.java:
```diff
@@ -32,7 +33,7 @@
     private final NotationParser<CharSequence, ?> notationParser;
 
-    public AbstractOptionElement(String optionName, Option option, Class<?> optionType, Class<?> declaringClass, NotationParser<CharSequence, ?> notationParser) {
+    AbstractOptionElement(String optionName, Option option, Class<?> optionType, Class<?> declaringClass, NotationParser<CharSequence, ?> notationParser) {
```
What's the main reason for making things package visible and final ? Is it to prevent end users from accidentally using internal APIs?

wilkinsona replied:
Force of habit really. I always try to minimise visibility and mark anything with only private constructors as final as it can't be sub-classes anyway. I can revert those changes if it doesn't fit with Gradle's code style

bmuschko replied:
Thanks for the info. We usually don't do that. Let's revert those changes for now. It will also make it easier to review the changes and get them into the ready-state.

bmuschko commented:
@wilkinsona Hi Andy, were you able to make any further progress on this PR? I'd love to have the feature in Gradle core.

wilkinsona commented:
Not yet. It's on my TODO list for this week but I need to take care of some other stuff first.

bmuschko commented:
Awesome! Looking forward to your changes.

wilkinsona commented:
I've made a little bit of progress and force-pushed an updated commit. The latest changes are:

- Reverted the changes to visibility and finality
- Improved the javadoc of `@Option` and `@OptionValues`
- Added `@Incubating` to `@Option` and `@OptionValues`

The options support seems to be fairly well covered by the integration tests in `TaskCommandLineConfigurationIntegrationSpec`. One thing that's missing is any use of `@OptionValues`. Is there anything else that you'd particularly like to see?

I'll tackle any necessary integration test updates and the documentation next. Unfortunately, that won't be till towards the end of August as I'm on holiday for the next couple of weeks.

bmuschko commented:
Thinking about it more I think we'll want to keep the internal classes instead of moving them. Those classes are internal but I know a bunch of plugin that use them. We'd automatically break them but just shuffling things around. Could you leave the "old" classes in place and add deprecation warnings for them? A message should be rendered that indicates that they will be remove in Gradle 5.0. In the Javadocs we'll want to point users to the new public API. All Gradle core plugins can starting using the public API with this change.

> I'll tackle any necessary integration test updates and the documentation next. Unfortunately, that won't be till towards the end of August as I'm on holiday for the next couple of weeks.

Perfect, thanks.

bmuschko commented:
> Unfortunately, that won't be till towards the end of August as I'm on holiday for the next couple of weeks.

Hi, @wilkinsona. Hope you had a nice vacation. Is this something you are planning to pick up again soon?

wilkinsona commented:
I did. Thanks.

I'd hoped to pick this up sooner, but Spring Boot M4 has rather got in the way. Sorry for the delay. All being well, I'll be able to look at this again next week.

wilkinsona commented:
With apologies for the delay, I've force-pushed a new commit that moves things forward. The changes since the last review are:

- Internal `@Option` has been reinstated and marked as deprecated
- Internal `@OptionValues` has been reinstated and marked as deprecated
- `OptionReader` looks for both internal and public variants of `@Option` and `@OptionValues`, logging a deprecation warning when the internal annotations are found
- Tests have been updated to verify that the internal annotations continue to work
- Integration tests have been updated to verify that the internal annotations log deprecation warnings

I think (hope) that the documentation updates are the final thing to tackle. Can you give me some pointers on how and where to make those updates please?

bmuschko commented:
> With apologies for the delay, I've force-pushed a new commit that moves things forward.

Thanks for the updated files. We'll have a look at the changes again soon.

> I think (hope) that the documentation updates are the final thing to tackle. Can you give me some pointers on how and where to make those updates please?

At the moment the options are just task-specific. I'd probably add the documentation to `tasks.adoc`. You will find that our documentation embeds sample code. We should do the same thing here. An important part of the documentation for this feature would be to list the limitations.

wilkinsona commented:
> At the moment the options are just task-specific. I'd probably add the documentation to tasks.adoc in the docs module. You will find that our documentation embeds sample code. We should do the same thing here. An important part of the documentation for this feature would be to list the limitations.

Can you offer some more guidance please, or perhaps take care of the documentation updates? I only have limited time left to spend on this and can't afford to spend time writing documentation that doesn't meet your unstated expectations.

bmuschko commented:
> Can you offer some more guidance please, or perhaps take care of the documentation updates? I only have limited time left to spend on this and can't afford to spend time writing documentation that doesn't meet your unstated expectations.

Sure, I can give you some more guidance. It's not 100% clear to me what information you are looking for. Are you asking about the type of content, the structure, how to organize? Here's what I would expect in general:

- Describes the purpose of the new API (what problem does it solve)
- Demonstrates the features of the API by example (the example code would be embedded into the text) e.g. https://github.com/gradle/gradle/issues/2402#issue-241067601
- Explains limitations or what does not work yet e.g. https://github.com/gradle/gradle/issues/2402#issuecomment-313514094

wilkinsona commented:
Unfortunately, following Spring Boot's move back to subclassing `JavaExec` for its `BootRun` task, I no longer have a requirement motivating this change. As such, I can't justify spending any more time working on it. Sorry. I will leave it to the Gradle team to decide what should happen to the proposed changes. FWIW, I am happy for someone else to build upon my proposed changes if they wish to do so.

bmuschko commented:
Thanks for all your work, @wilkinsona. I finished off your PR in a new one: https://github.com/gradle/gradle/pull/4040/.
